tools/bpm.c

```diff
@@ -0,0 +1,176 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <string.h>
+
+#include "../src/msg.h"
+#include "../src/acq/channel.h"
+
+#include "msg.h"
+#include "sig.h"
+#include "util.h"
+
+struct channel_data {
+	uint8_t channel;
+	uint16_t frequency;
+	uint64_t sample_index;
+	uint64_t old_peak_index;
+	uint32_t old_peak_height;
+	uint64_t new_peak_index;
+	uint32_t new_peak_height;
+	bool in_peak;
+};
+
+uint32_t DEFAULT_PEAK_THRESHOLD = 3 * (UINT32_MAX / 4); // half-way above the middle.
+
+static inline double calculate_bpm(const struct channel_data *channel)
+{
+	return 60.0 * (double) channel->frequency /
+		(double) (channel->new_peak_index - channel->old_peak_index);
+}
+
+void channel_process_sample(struct channel_data *channel,
+		uint32_t peak_threshold, const bl_msg_sample_data_t *msg)
+{
+	uint32_t value;
+	for(uint8_t i = 0; i < msg->count; i++) {
+		if (msg->type == BL_MSG_SAMPLE_DATA16) {
+			/* Upscale 16-bit samples to 32-bit. */
+			value = msg->data16[i];
+			value = (value << 16) | value;
+		} else {
+			value = msg->data32[i];
+		}
+
+		if (value >= peak_threshold) {
+			if (!channel->in_peak) {
+				// Entered a new peak
+				channel->in_peak = true;
+			}
+			if (value > channel->new_peak_height) {
+				channel->new_peak_height = value;
+				channel->new_peak_index = channel->sample_index;
+			}
+
+		} else if (channel->in_peak) {
+			// Just left a peak
+			channel->in_peak = false;
+			if (channel->old_peak_height) {
+				// a previous peak exists to compare times for
+				printf("%u,%u\n", channel->channel,
+						(unsigned) calculate_bpm(channel));
+			}
+			channel->old_peak_index = channel->new_peak_index;
+			channel->old_peak_height = channel->new_peak_height;
+			channel->new_peak_index = 0;
+			channel->new_peak_height = 0;
+		}
+		channel->sample_index++;
+	}
+}
+
+void channel_start(struct channel_data *channel,
+		const bl_msg_start_t *msg)
+{
+	channel->frequency = msg->frequency;
+}
+
+void init_channel(struct channel_data *channel,
+		const bl_msg_channel_conf_t *msg)
+{
+	channel->channel = msg->channel;
+}
+
+int read_stream(uint32_t peak_threshold)
+{
+	union bl_msg_data msg; // message for reading into
+	struct channel_data channels[BL_ACQ_CHANNEL_COUNT] = {0};
+	struct channel_data *channel;
+	while (!bl_sig_killed && bl_msg_yaml_parse(stdin, &msg)) {
+		switch(msg.type) {
+		case BL_MSG_CHANNEL_CONF:
+			init_channel(channels + msg.channel_conf.channel,
+					&msg.channel_conf);
```

Comment:
Since I've written a lot of these tools, and they're all very similar, I've been working towards having a consistent way of processing the stream. Hence I'd rather have some redundancy now.

Reply:
Ah, cool, good idea.  :smile_cat: